### PR TITLE
feat(communities): Makes `requestToJoinCommunity` async

### PR DESCRIPTION
**Pull request by borismelnik**
Fixes: #10719

### What does the PR do

- Makes request to join async
- Add `Loading...` everywhere when `Pending` state was used

### Affected areas

Request to join community 

### StatusQ checklist

- [ ] add documentation if necessary (new component, new feature)
- [ ] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- [ ] test changes in both light and dark theme?

### CODE REVIEW

File: src/app_service/service/community/service.nim
Comment (jrainville) [inline src/app_service/service/community/service.nim:1397]: Do we really want to remove the request to join when we get the answer? I think it most cases, this will go in "pending" state, so we don't want to delete it, just change the state?

Maybe we just need to change the event from `DELETED` to `UPDATED` so that it can work for all the state changes `loading -> pending -> accepted/rejected`

**Review by jrainville — COMMENTED** [review]
Nice work. I have a question about the removal of the request

File: src/app_service/service/community/service.nim
Comment (borismelnik) [reply src/app_service/service/community/service.nim:1397]: I thought about it, but request there is kind a `fake` cos in database it's not exists. Technically you may be right, will fix it.

Comment (borismelnik) [reply src/app_service/service/community/service.nim]: Updated :)

Comment (borismelnik) [reply src/app_service/service/community/service.nim]: Delete method still make sense as a clean up method

File: src/app_service/service/community/service.nim
Comment (0x-r4bbit) [inline src/app_service/service/community/service.nim:1380]: So this part looks a bit backwards to me. Seems like we're optimistically updating our locally known requests.

Does anything speak against:

1. Make request async
2. Wait for callback when request was made (just like we do with waiting for synchronous call to be done)
3. When callback doesn't contain an error, we know it was successful, **then** emit ADDED event
4. In case of error, emit error event so we can show a notification accordingly

File: src/app_service/service/community/service.nim
Comment (0x-r4bbit) [inline src/app_service/service/community/service.nim:1409]: Here's where I think we should emit `ADDED` event (or `UPDATED`, but probably `ADDED` because we're only adding membership requests, we're never really updating an existing one

File: src/app_service/service/community/service.nim
Comment (0x-r4bbit) [inline src/app_service/service/community/service.nim:1720]: I don't think we need the state check for `Loading` here.
As mentioned above, I don't think we need the loading state as a type at all.

Loading state can be handled entirely in the view. The idea would be to

1. Make request (set loading state to true)
2 When request finished, set loading state to false (now the request is pending)
3. If request fails due to an error, set loading state to false (ideally also show a notification that it failed due to some reason)

File: ui/app/AppLayouts/Chat/views/CommunityColumnView.qml
Comment (0x-r4bbit) [inline ui/app/AppLayouts/Chat/views/CommunityColumnView.qml:84]: When it's loading I don't think we want to show any text here. We just put the button in loading state.

**Review by 0x-r4bbit — COMMENTED** [review]
Great work!
I just think the can simplify this a bit.

The new `RequestToJoinType.Loading` variant is IMO not necessary, unless we expect a user to perform requests to join for multiple communities simultaneously, that will take a long time to process.

I think in most of the cases there'll be only one request to join at a time

File: src/app_service/service/community/dto/community.nim
Comment (borismelnik) [reply src/app_service/service/community/dto/community.nim:13]: No, cos `Pending` means we sent request and waiting for approve from owner side. On this state we can cancel it if we want.

Comment (0x-r4bbit) [reply src/app_service/service/community/dto/community.nim:13]: Realized that `Loading` is meant for UI only, so it's not the same as `Pending` but it also doesn't belong here IMO

**Comment by borismelnik** [comment]
@0x-r4bbit as I undestand, you mean that after click on `RequestToJoin` we put the button in loading state and when async request was done - we just change it? It's can't take a lot of time right? We not expect that time to async request is enough to change section and back?

**Comment by borismelnik** [comment]
> > @0x-r4bbit as I undestand, you mean that after click on RequestToJoin we put the button in loading state and when async request was done - we just change it? It's can't take a lot of time right? We not expect that time to async request is enough to change section and back?
> 
> Theoretically it could take a long time, depending on how many permissions are attached to the community and how fast the RPC endpoint is to return data.
> 
> But in practice I don't think this taking longer than a couple of seconds.

so if we not planning to make it bigger then your comments makes sense, otherwise the Loading state is necessary :)

**Comment by 0x-r4bbit** [comment]
> so if we not planning to make it bigger then your comments makes sense, otherwise the Loading state is necessary :)

Yea, I guess if we want to be 100% correct, that we'll have to track loading state of every ongoing request. 
I'm gonna leave this be your call! :)

**Comment by borismelnik** [comment]
> > so if we not planning to make it bigger then your comments makes sense, otherwise the Loading state is necessary :)
> 
> Yea, I guess if we want to be 100% correct, that we'll have to track loading state of every ongoing request. I'm gonna leave this be your call! :)

anyway, lets wait for another opinion from @jrainville or @osmaczko :)

**Comment by borismelnik** [comment]
one thing which I don't like in this PR its a custom request type just for us app.

**Comment by 0x-r4bbit** [comment]
> one thing which I don't like in this PR its a custom request type just for us app.

I agree. IMO it does not belong there. It's purely UI related and considered temporary. So better to just keep track of it in the view layer.

**Comment by jrainville** [comment]
@borismelnik I agree with @0x-r4bbit 's points. It can be only UI side.

The loading shouldn't ever be long enough that the user has time to go to another section and comes back. The goal is just to have a visual indicator that something is going on, without freezing, so the user doesn't try to click the button again.

**Comment by borismelnik** [comment]
> @borismelnik I agree with @0x-r4bbit 's points. It can be only UI side.
> 
> The loading shouldn't ever be long enough that the user has time to go to another section and comes back. The goal is just to have a visual indicator that something is going on, without freezing, so the user doesn't try to click the button again.

Great, lets do it :)

**Comment by borismelnik** [comment]
@0x-r4bbit @jrainville Updated :) Sorry guys, I thought it should more complex. Now it looks much better :)

**Review by 0x-r4bbit — APPROVED** [review]

File: src/app_service/service/community/service.nim
Comment (osmaczko) [inline src/app_service/service/community/service.nim:143]: Seems unused

**Comment by osmaczko** [comment]
> > > so if we not planning to make it bigger then your comments makes sense, otherwise the Loading state is necessary :)
> > 
> > 
> > Yea, I guess if we want to be 100% correct, that we'll have to track loading state of every ongoing request. I'm gonna leave this be your call! :)
> 
> anyway, lets wait for another opinion from @jrainville or @osmaczko :)

I am okay with both approaches. Preferring simplicity over a complicated solution when the probability of an uncovered case is very low will help make the code easier to maintain. On the other hand, the code is not self-adaptive to potential changes in the duration of the requests and may be sensitive to issues such as poor internet connection.

**Review by osmaczko — APPROVED** [review]

File: src/app_service/service/community/service.nim
Comment (borismelnik) [reply src/app_service/service/community/service.nim:143]: oh thanks I missed it

**Pull request merged**
